File: node-bindings/classic-node/__test__/update_services_conformance.ts
import * as classic from "../index.js";
import { mkdtempSync, mkdirSync, readdirSync, readFileSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
type JsonObject = Record<string, any>;

/** Execute notification checks with the real native binding and one isolated cache. */
export async function observeUpdateServices(fixture: JsonObject, scenarioId: string): Promise<JsonObject> {
  const service = process.env.CLASSIC_CONFORMANCE_SERVICE;
  if (!service) throw new Error("CLASSIC_CONFORMANCE_SERVICE is required");
  const base = `${service.replace(/\/$/, "")}/${scenarioId}`;
  const config = JSON.stringify({ github_api_base_url: `${base}/api`, notification_pages_url: `${base}/pages`, timeout_ms: fixture.timeoutMs });
  const root = mkdtempSync(join(tmpdir(), "classic-update-"));
  const cache = join(root, "cache");
  mkdirSync(cache);
  try {
    const results: JsonObject[] = [];
    for (let check = 0; check < fixture.checks; check++) {
      try {
        const status = await classic.checkAppNotificationConfigured("conformance", "updates", fixture.installedVersion, config, cache);
        results.push({ status: {
          classification: status.classification.replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`),
          latestVersion: status.latestVersion, publishedAt: status.publishedAt,
          minSupportedVersion: status.minSupportedVersion ?? null,
          display: status.display ? { title: status.display.title, body: status.display.body, ctaUrl: status.display.ctaUrl ?? null } : null,
          parseError: status.parseError ?? null,
        }, error: null });
      } catch (error) {
        if (!(error instanceof Error)) throw error;
        let code: string;
        if (error.message.startsWith("FETCH_FAILED:")) code = "fetch_failed";
        else if (error.message.startsWith("DECODE:")) code = "decode";
        else if (error.message.startsWith("INSTALLED_VERSION_PARSE:")) code = "installed_version";
        else if (error.message.startsWith("UPDATE_ERROR: manifest_version ") && error.message.includes(" not supported ")) code = "unsupported_version";
        else throw error;
        results.push({ status: null, error: { code } });
      }
    }
    const files = readdirSync(cache).sort().map(name => ({ path: `cache/${name}`, hex: readFileSync(join(cache, name)).toString("hex") }));
    return { results, files };
  } finally {
    rmSync(root, { recursive: true, force: true });
  }
}

File: python-bindings/tests/update_services_conformance.py
"""Observe notification transport and cache bytes through Python bindings."""

import json
import os
import re
import tempfile
from pathlib import Path
from typing import Any


def observe_update_services(fixture: dict[str, Any], scenario_id: str) -> dict[str, Any]:
    """Run repeated checks against the controlled service with one isolated cache."""
    import classic_update

    base = os.environ["CLASSIC_CONFORMANCE_SERVICE"].rstrip("/") + "/" + scenario_id
    config = json.dumps({"github_api_base_url": base + "/api", "notification_pages_url": base + "/pages", "timeout_ms": fixture["timeoutMs"]})
    with tempfile.TemporaryDirectory(prefix="classic-update-") as directory:
        cache = Path(directory) / "cache"
        cache.mkdir()
        results = []
        for _ in range(fixture["checks"]):
            try:
                status = classic_update.check_app_notification_configured("conformance", "updates", fixture["installedVersion"], config, str(cache))
                display = status.display
                results.append({"status": {
                    "classification": re.sub(r"(?<!^)(?=[A-Z])", "_", status.classification).lower(),
                    "latestVersion": status.latest_version, "publishedAt": status.published_at,
                    "minSupportedVersion": status.min_supported_version,
                    "display": None if display is None else {"title": display.title, "body": display.body, "ctaUrl": display.cta_url},
                    "parseError": status.parse_error,
                }, "error": None})
            except classic_update.ClassicNotificationError as error:
                if isinstance(error, classic_update.ClassicNotificationFetchFailed):
                    code = "fetch_failed"
                elif isinstance(error, classic_update.ClassicNotificationDecodeError):
                    code = "decode"
                elif isinstance(error, classic_update.ClassicNotificationInstalledVersionParseError):
                    code = "installed_version"
                elif str(error).startswith("manifest_version ") and " not supported " in str(error):
                    code = "unsupported_version"
                else:
                    raise
                results.append({"status": None, "error": {"code": code}})
        files = [{"path": path.relative_to(directory).as_posix(), "hex": path.read_bytes().hex()} for path in sorted(cache.rglob("*")) if path.is_file()]
        return {"results": results, "files": files}
